wasm/validation.cpp

```diff
@@ -0,0 +1,373 @@
+
+using namespace wasm;
+using namespace wasm::instructions;
+
+using ValueOrUnknown = std::variant<std::monostate, ValueType>;
+
+// NOLINTNEXTLINE(cert-err58-cpp)
+std::map<ValidationError, std::string_view> const validation_error_str = {
+        {ValidationError::ControlStackEmpty, "Attempted to pop from the control stack, but the control stack is empty"},
+        {ValidationError::LabelInvalid, "Attempted to branch to a label which isn't valid"},
+        {ValidationError::LocalUndefined, "Attempted to index a local which isn't defined in the current code entry"},
+        {ValidationError::MemoryBadAlignment, "Attempted a load or store with a bad alignment value"},
+        {ValidationError::MemoryEmpty, "Attempted a load, but memory is empty"},
+        {ValidationError::MemorySectionUndefined, "Attempted a load or store, but no memory section was defined"},
+        {ValidationError::TypeSectionUndefined,
+                "Encountered a control instruction with a type use, but no type section was defined"},
+        {ValidationError::TypeUndefined, "Attempted to index a type which isn't defined in the type section"},
+        {ValidationError::UnknownInstruction, "Unknown instruction encountered"},
+        {ValidationError::ValueStackHeightMismatch,
+                "Value stack height on exiting a control frame does not match the height on entry"},
+        {ValidationError::ValueStackUnderflow, "Attempted to pop from value stack, but stack height would underflow"},
+        {ValidationError::ValueStackUnexpected,
+                "Attempted to pop expected value from value stack, but got unexpected value"},
+};
+
+std::string_view to_string(ValidationError err) {
+    return validation_error_str.at(err);
+}
+
+struct ControlFrame {
+    Instruction i;
+    std::vector<ValueType> params;
+    std::vector<ValueType> results;
+    std::size_t stack_height;
+    bool unreachable;
+};
+
+namespace {
+
+void push_val(std::vector<ValueOrUnknown> &value_stack, ValueOrUnknown const &val) {
+    value_stack.push_back(val);
+}
+
+tl::expected<ValueOrUnknown, ValidationError> pop_val(
+        std::vector<ValueOrUnknown> &value_stack, std::vector<ControlFrame> &control_stack) {
+    // This check assumes that there's always a frame on the control stack.
+    if (value_stack.size() == control_stack.back().stack_height && control_stack.back().unreachable) {
+        return std::monostate{};
+    }
+
+    if (value_stack.size() == control_stack.back().stack_height) {
+        return tl::unexpected{ValidationError::ValueStackUnderflow};
+    }
+
+    auto val = value_stack.back();
+
+    value_stack.pop_back();
+
+    return val;
+}
+
+tl::expected<ValueOrUnknown, ValidationError> pop_val_expect(std::vector<ValueOrUnknown> &value_stack,
+        std::vector<ControlFrame> &control_stack,
+        ValueOrUnknown const &expected) {
+    auto actual = pop_val(value_stack, control_stack);
+
+    // propagate pop_val error upward
```

Comment:
No need for comments like this. Error propagation with std::expected isn't unexpected or anything.